File: scripts/generate-smart-previews.py
"""
generate-smart-previews.py
==========================
Génère des previews intelligentes pour les documents ACER :
- Sélectionne la première page avec des photos (images embarquées)
- Ignore les premières pages (couverture, sécurité, sommaire)
- Upload vers Supabase Storage logos/previews/{slug}.jpg (upsert)

Usage :
    python scripts/generate-smart-previews.py [--brand ACER] [--dry-run]

Dépendances : pip install pymupdf python-dotenv requests
"""

import os
import sys
import json
import fitz  # PyMuPDF
import requests

# ── Config ────────────────────────────────────────────────────────────────────
REPORT_PATH    = r"C:\Users\adm\Claude Doc GB test\service-manuals-pro\scripts\docs-a-classer-report-informatique.json"
DOCS_EN_LIGNE  = r"C:\Users\adm\Documents\SHEMATHEQUE\DOCS EN LIGNE\Informatique"
STORAGE_BUCKET = "logos"
ENV_PATH       = r"C:\Users\adm\Claude Doc GB test\service-manuals-pro\.env.local"

# Pages à ignorer en début de document (couverture, sécurité, sommaire)
SKIP_FIRST_N   = 2   # ignorer pages 0 et 1 (index), commencer à chercher à partir de page index 2
# Si aucune page image trouvée dans les N premières pages, on abandonne la recherche
MAX_SCAN_PAGES = 50
# Seuil minimum d'images pour considérer une page comme "riche"
MIN_IMAGES     = 1

# ── Args ──────────────────────────────────────────────────────────────────────
args     = sys.argv[1:]
DRY_RUN  = "--dry-run" in args
BRAND_FILTER = None
if "--brand" in args:
    idx = args.index("--brand")
    if idx + 1 < len(args):
        BRAND_FILTER = args[idx + 1].upper()

# ── Load env (lecture manuelle pour éviter les problèmes de parsing dotenv) ───
def read_env_file(path):
    """Lit un fichier .env/.env.local et retourne un dict propre."""
    env = {}
    try:
        with open(path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                key, _, val = line.partition("=")
                key = key.strip()
                val = val.strip()
                # Retirer guillemets encadrants
                if len(val) >= 2 and val[0] in ('"', "'") and val[-1] == val[0]:
                    val = val[1:-1]
                env[key] = val
    except FileNotFoundError:
        print(f"ERREUR : fichier env introuvable : {path}")
        sys.exit(1)
    return env

env_vars    = read_env_file(ENV_PATH)
SUPABASE_URL = env_vars.get("NEXT_PUBLIC_SUPABASE_URL", "").strip()
SUPABASE_KEY = env_vars.get("SUPABASE_SERVICE_ROLE_KEY", "").strip()

if not SUPABASE_URL or not SUPABASE_KEY:
    print("ERREUR : variables NEXT_PUBLIC_SUPABASE_URL ou SUPABASE_SERVICE_ROLE_KEY manquantes")
    print(f"  Clés trouvées : {list(env_vars.keys())}")
    sys.exit(1)

# Affichage masqué pour vérification
print(f"SUPABASE_URL : {SUPABASE_URL}")
print(f"SERVICE_KEY  : {SUPABASE_KEY[:20]}...{SUPABASE_KEY[-10:]}")

# ── Load report ───────────────────────────────────────────────────────────────
with open(REPORT_PATH, encoding="utf-8") as f:
    report = json.load(f)

docs = report.get("docs", [])

# Filtrer : uniquement les docs importés de la marque cible
targets = []
for doc in docs:
    if doc.get("status") != "imported":
        continue
    brand = doc.get("brand", "").upper()
    if BRAND_FILTER and brand != BRAND_FILTER:
        continue
    slug = doc.get("slug")
    original_filename = os.path.basename(doc.get("original_filename", "") or doc.get("original_path", ""))
    if not slug or not original_filename:
        continue
    # Retrouver le PDF dans DOCS EN LIGNE
    brand_folder = os.path.join(DOCS_EN_LIGNE, brand)
    pdf_path = os.path.join(brand_folder, original_filename)
    targets.append({
        "slug": slug,
        "pdf_path": pdf_path,
        "brand": brand,
        "original_filename": original_filename,
    })

print(f"Documents cibles : {len(targets)}")
if BRAND_FILTER:
    print(f"Filtre marque    : {BRAND_FILTER}")
if DRY_RUN:
    print("MODE DRY-RUN — aucun upload")
print()


def find_best_page(doc_fitz):
    """
    Trouve la première page avec des images embarquées (photos de démontage).
    Stratégie :
      1. Chercher à partir de SKIP_FIRST_N une page avec >= MIN_IMAGES images
      2. Si non trouvé dans MAX_SCAN_PAGES, revenir à la page index 1 (ou 0 si 1 seule page)
    """
    n = len(doc_fitz)

    # Cherche une page riche en images à partir de SKIP_FIRST_N
    for i in range(SKIP_FIRST_N, min(n, SKIP_FIRST_N + MAX_SCAN_PAGES)):
        page = doc_fitz[i]
        images = page.get_images(full=False)
        if len(images) >= MIN_IMAGES:
            return i, len(images)

    # Fallback : page 1 (ou 0 si doc trop court)
    fallback = min(1, n - 1)
    return fallback, 0


def generate_thumbnail(doc_fitz, page_idx, width=800):
    """Génère un JPEG 800px de large depuis la page indiquée."""
    page = doc_fitz[page_idx]
    scale = width / page.rect.width
    mat = fitz.Matrix(scale, scale)
    pix = page.get_pixmap(matrix=mat, alpha=False)
    return pix.tobytes("jpeg")


def upload_preview(slug, jpg_bytes):
    """Upsert du JPEG dans Supabase Storage logos/previews/{slug}.jpg"""
    storage_path = f"previews/{slug}.jpg"
    url = f"{SUPABASE_URL}/storage/v1/object/{STORAGE_BUCKET}/{storage_path}"
    headers = {
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "image/jpeg",
        "x-upsert": "true",
    }
    r = requests.post(url, headers=headers, data=jpg_bytes, timeout=30)
    if r.status_code not in (200, 201):
        raise RuntimeError(f"HTTP {r.status_code} : {r.text[:200]}")
    return storage_path


# ── Traitement ────────────────────────────────────────────────────────────────
ok = 0
errors = 0
fallbacks = 0

for i, t in enumerate(targets):
    slug      = t["slug"]
    pdf_path  = t["pdf_path"]
    brand     = t["brand"]

    if not os.path.exists(pdf_path):
        print(f"  [{i+1}/{len(targets)}] ✗ INTROUVABLE : {t['original_filename']}")
        errors += 1
        continue

    try:
        doc_fitz = fitz.open(pdf_path)
        page_idx, n_images = find_best_page(doc_fitz)

        # Indicator : fallback ou page trouvée ?
        if n_images == 0:
            marker = f"[fallback p.{page_idx+1}]"
            fallbacks += 1
        else:
            marker = f"[p.{page_idx+1}, {n_images} img]"

        if DRY_RUN:
            print(f"  [{i+1}/{len(targets)}] DRY  {marker}  {slug}")
            doc_fitz.close()
            ok += 1
            continue

        jpg_bytes = generate_thumbnail(doc_fitz, page_idx)
        doc_fitz.close()

        upload_preview(slug, jpg_bytes)
        print(f"  [{i+1}/{len(targets)}] ✓ {marker}  {slug}")
        ok += 1

    except Exception as e:
        print(f"  [{i+1}/{len(targets)}] ✗ ERREUR {slug} : {e}")
        errors += 1

print()
print(f"Terminé — {ok} OK | {fallbacks} fallback | {errors} erreurs")

File: scripts/generate-smart-previews.mjs
/**
 * generate-smart-previews.mjs
 * ============================
 * Génère des previews intelligentes pour les documents d'une marque :
 * - Sélectionne la première page avec des images (photos de démontage)
 * - Ignore les premières pages (couverture, sécurité, sommaire)
 * - Upload vers Supabase Storage logos/previews/{slug}.jpg (upsert)
 *
 * Usage :
 *   node scripts/generate-smart-previews.mjs <Brand> [--dry-run]
 *   node scripts/generate-smart-previews.mjs ACER --dry-run
 *   node scripts/generate-smart-previews.mjs ACER
 */

import dotenv from 'dotenv';
dotenv.config({ path: '.env.local' });

import { createClient } from '@supabase/supabase-js';
import { createRequire } from 'module';
import fs   from 'fs';
import path from 'path';

const require = createRequire(import.meta.url);
const pdfjs  = require('pdfjs-dist');
const canvas = require('canvas');

const STANDARD_FONTS = 'C:/Users/adm/Claude Doc GB test/service-manuals-pro/node_modules/pdfjs-dist/standard_fonts/';

// ── Config ────────────────────────────────────────────────────────────────────
const REPORT_PATH   = path.join('scripts', 'docs-a-classer-report-informatique.json');
const DOCS_EN_LIGNE = 'C:\\Users\\adm\\Documents\\SHEMATHEQUE\\DOCS EN LIGNE\\Informatique';
const STORAGE_BUCKET = 'logos';

// Pages à ignorer en tête de document
const SKIP_FIRST_N  = 2;    // commencer à chercher à partir de l'index 2 (page 3)
const MAX_SCAN      = 50;   // scanner au plus 50 pages
const MIN_IMAGES    = 1;    // seuil minimum d'images pour retenir la page
const THUMB_WIDTH   = 800;  // largeur thumbnail en pixels

// ── Args ──────────────────────────────────────────────────────────────────────
const args      = process.argv.slice(2);
const DRY_RUN   = args.includes('--dry-run');
const BRAND_ARG = args.find(a => !a.startsWith('--'));

if (!BRAND_ARG) {
  console.error('Usage : node scripts/generate-smart-previews.mjs <Brand> [--dry-run]');
  process.exit(1);
}
const BRAND_FILTER = BRAND_ARG.toUpperCase();

// ── Supabase ──────────────────────────────────────────────────────────────────
const supabase = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY
);

// ── Load report ───────────────────────────────────────────────────────────────
if (!fs.existsSync(REPORT_PATH)) {
  console.error(`Rapport introuvable : ${REPORT_PATH}`);
  process.exit(1);
}
const report = JSON.parse(fs.readFileSync(REPORT_PATH, 'utf-8'));
const docs   = report.docs || [];

// Filtrer les docs importés de la marque cible
const targets = docs
  .filter(d => d.status === 'imported' && (d.brand || '').toUpperCase() === BRAND_FILTER)
  .map(d => {
    const originalFilename = path.basename(d.original_filename || d.original_path || '');
    const brandFolder      = path.join(DOCS_EN_LIGNE, BRAND_FILTER);
    const pdfPath          = path.join(brandFolder, originalFilename);
    return { slug: d.slug, pdfPath, originalFilename };
  })
  .filter(d => d.slug && d.originalFilename);

console.log(`Marque          : ${BRAND_FILTER}`);
console.log(`Documents cibles: ${targets.length}`);
if (DRY_RUN) console.log('MODE DRY-RUN — aucun upload');
console.log();

// ── Helpers pdfjs ─────────────────────────────────────────────────────────────

/**
 * Compte les opérations image sur une page pdfjs.
 * paintImageXObject = image embarquée (photo, schéma)
 */
async function countImagesOnPage(page) {
  try {
    const ops = await page.getOperatorList();
    return ops.fnArray.filter(fn =>
      fn === pdfjs.OPS.paintImageXObject ||
      fn === pdfjs.OPS.paintInlineImageXObject ||
      fn === pdfjs.OPS.paintImageMaskXObject
    ).length;
  } catch {
    return 0;
  }
}

/**
 * Trouve la meilleure page (la plus riche en images, à partir de SKIP_FIRST_N).
 * Retourne { pageIdx, nImages }
 */
async function findBestPage(pdfjsDoc) {
  const n = pdfjsDoc.numPages;

  for (let i = SKIP_FIRST_N; i < Math.min(n, SKIP_FIRST_N + MAX_SCAN); i++) {
    const page   = await pdfjsDoc.getPage(i + 1); // pdfjs: 1-based
    const nImg   = await countImagesOnPage(page);
    page.cleanup();
    if (nImg >= MIN_IMAGES) {
      return { pageIdx: i, nImages: nImg };
    }
  }

  // Fallback : page 1 (index 1) ou 0 si doc très court
  const fallback = Math.min(1, n - 1);
  return { pageIdx: fallback, nImages: 0 };
}

/**
 * Rend une page en JPEG Buffer (width = THUMB_WIDTH px).
 */
async function renderPage(pdfjsDoc, pageIdx) {
  const page     = await pdfjsDoc.getPage(pageIdx + 1); // 1-based
  const viewport = page.getViewport({ scale: 1 });
  const scale    = THUMB_WIDTH / viewport.width;
  const vp       = page.getViewport({ scale });

  const canvasEl  = canvas.createCanvas(Math.round(vp.width), Math.round(vp.height));
  const ctx       = canvasEl.getContext('2d');

  await page.render({
    canvasContext: ctx,
    viewport:      vp,
    canvasFactory: {
      create(w, h) { const c = canvas.createCanvas(w, h); return { canvas: c, context: c.getContext('2d') }; },
      reset(obj, w, h) { obj.canvas.width = w; obj.canvas.height = h; },
      destroy(obj) {},
    },
  }).promise;

  page.cleanup();
  return canvasEl.toBuffer('image/jpeg', { quality: 0.85 });
}

/**
 * Upload vers Supabase Storage (upsert).
 */
async function uploadPreview(slug, jpgBuffer) {
  const storagePath = `previews/${slug}.jpg`;
  const { error } = await supabase.storage
    .from(STORAGE_BUCKET)
    .upload(storagePath, jpgBuffer, { contentType: 'image/jpeg', upsert: true });
  if (error) throw new Error(`Supabase upload error: ${error.message}`);
  return storagePath;
}

// ── Boucle principale ─────────────────────────────────────────────────────────
let ok = 0, errors = 0, fallbacks = 0;

for (let i = 0; i < targets.length; i++) {
  const { slug, pdfPath, originalFilename } = targets[i];
  const num = `[${i + 1}/${targets.length}]`;

  if (!fs.existsSync(pdfPath)) {
    console.log(`  ${num} ✗ INTROUVABLE : ${originalFilename}`);
    errors++;
    continue;
  }

  try {
    const data = new Uint8Array(fs.readFileSync(pdfPath));
    const pdfjsDoc = await pdfjs.getDocument({
      data,
      standardFontDataUrl: STANDARD_FONTS,
      verbosity: 0,
    }).promise;

    const { pageIdx, nImages } = await findBestPage(pdfjsDoc);
    const marker = nImages === 0
      ? `[fallback p.${pageIdx + 1}]`
      : `[p.${pageIdx + 1}, ${nImages} img]`;

    if (nImages === 0) fallbacks++;

    if (DRY_RUN) {
      console.log(`  ${num} DRY  ${marker}  ${slug}`);
      pdfjsDoc.destroy();
      ok++;
      continue;
    }

    const jpgBuffer = await renderPage(pdfjsDoc, pageIdx);
    pdfjsDoc.destroy();

    await uploadPreview(slug, jpgBuffer);
    console.log(`  ${num} ✓ ${marker}  ${slug}`);
    ok++;

  } catch (err) {
    console.log(`  ${num} ✗ ERREUR ${slug} : ${err.message}`);
    errors++;
  }
}

console.log();
console.log(`Terminé — ${ok} OK | ${fallbacks} fallback | ${errors} erreurs`);
